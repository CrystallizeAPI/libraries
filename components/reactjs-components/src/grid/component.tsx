export * from './CSSGrid';
export * from './Table';
export * from './RowCol';
