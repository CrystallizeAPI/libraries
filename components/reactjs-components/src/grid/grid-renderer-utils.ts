import { GridCell, GridDimensions } from './types';

export const getGridDimensions = (rows: any[]): GridDimensions => {
    return {
        rows: rows.length,
        cols: rows[0].columns.reduce((acc: number, col: any) => acc + col.layout.colspan, 0),
    };
};

export function getPositionnableCellClassNames(cell: GridCell, { rows, cols }: GridDimensions): string {
    const { colspan, rowspan, rowIndex, colIndex } = cell.layout;
    const isLastRow = rowIndex + rowspan === rows;
    const isLastCol = colIndex + colspan === cols;

    return `cell-${rowIndex}-${colIndex} ${rowIndex == 0 ? 'first-row' : ''} ${colIndex == 0 ? 'first-col' : ''} ${
        isLastRow ? 'last-row' : ''
    } ${isLastCol ? 'last-col' : ''}`.replace(/\s+/g, ' ');
}

export function getPositionnablRowClassNames({ rowIndex }: { rowIndex: number }, { rows }: GridDimensions): string {
    return `row-${rowIndex} ${rowIndex == 0 ? 'first-row' : ''} ${rowIndex == rows - 1 ? 'last-row' : ''}`.replace(
        /\s+/g,
        ' ',
    );
}
