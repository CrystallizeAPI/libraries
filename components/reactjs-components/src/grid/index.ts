export * from './GridRenderer';
export * from './component';
export * from './types';
export * from './grid-renderer-utils';
