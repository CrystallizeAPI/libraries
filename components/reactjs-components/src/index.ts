export * from './image/index';
export * from './grid/index';
export * from './content-transformer/index';
export * from './video/index';
