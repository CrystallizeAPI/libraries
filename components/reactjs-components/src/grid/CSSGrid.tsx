import { FunctionComponent } from 'react';
import { getPositionnableCellClassNames } from './grid-renderer-utils';
import { CSSGridProps, GridCell } from './types';

export const CSSGrid: FunctionComponent<CSSGridProps> = ({
    cellComponent,
    cells,
    children,
    dimensions,
    style,
    styleForCell,
    ...props
}) => {
    const CellComponent = cellComponent;
    return (
        <div
            style={{
                display: 'grid',
                gridTemplateColumns: `repeat(${dimensions.cols}, 1fr)`,
                ...style,
            }}
            className="crystallize-grid crystallize-grid--css-grid"
            {...props}
        >
            {children && children({ cells, dimensions })}
            {!children &&
                cells.map((cell: GridCell, i: number) => {
                    const defaultStyles = {
                        gridColumn: `span ${cell.layout.colspan}`,
                        gridRow: `span ${cell.layout.rowspan}`,
                    };
                    const cellStyles = styleForCell
                        ? styleForCell(cell, defaultStyles) || defaultStyles
                        : defaultStyles;
                    return (
                        <div
                            key={`cell-${i}`}
                            className={`crystallize-grid__cell ${getPositionnableCellClassNames(cell, dimensions)}`}
                            style={cellStyles}
                        >
                            <CellComponent cell={cell} dimensions={dimensions} />
                        </div>
                    );
                })}
        </div>
    );
};
