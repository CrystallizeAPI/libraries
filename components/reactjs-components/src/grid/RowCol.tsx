import { FunctionComponent } from 'react';
import { getPositionnableCellClassNames, getPositionnablRowClassNames } from './grid-renderer-utils';
import { GridRow, GridCell, RowColGridProps } from './types';

export const RowCol: FunctionComponent<RowColGridProps> = ({
    cellComponent,
    grid,
    children,
    dimensions,
    styleForCell,
    ...props
}) => {
    const CellComponent = cellComponent;
    return (
        <div className="crystallize-grid crystallize-row-col-table" {...props}>
            {children && children({ grid, dimensions })}
            {!children &&
                grid.map((row: GridRow, rowIndex: number) => {
                    return (
                        <div
                            key={`row-${rowIndex}`}
                            className={`crystallize-grid-row row ${getPositionnablRowClassNames(
                                { rowIndex },
                                dimensions,
                            )}`}
                        >
                            {row.columns.map((cell: GridCell, cellIndex: number) => {
                                const cellStyles = styleForCell ? styleForCell(cell, {}) || {} : {};
                                const classes = getPositionnableCellClassNames(cell, dimensions);
                                return (
                                    <div
                                        key={`cell-${rowIndex}-${cellIndex}`}
                                        className={`crystallize-grid__cell col ${classes}`}
                                        style={cellStyles}
                                    >
                                        <CellComponent cell={cell} dimensions={dimensions} />
                                    </div>
                                );
                            })}
                        </div>
                    );
                })}
        </div>
    );
};
