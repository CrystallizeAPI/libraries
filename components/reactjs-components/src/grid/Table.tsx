import { FunctionComponent } from 'react';
import { getPositionnableCellClassNames, getPositionnablRowClassNames } from './grid-renderer-utils';
import { GridCell, GridRow, TableGridProps } from './types';

export const Table: FunctionComponent<TableGridProps> = ({
    cellComponent,
    grid,
    children,
    dimensions,
    styleForCell,
    ...props
}) => {
    const CellComponent = cellComponent;
    return (
        <table className="crystallize-grid crystallize-grid--table" {...props}>
            <thead>
                <tr>
                    {new Array(dimensions.cols).fill(0).map((v, i) => (
                        <th key={`th-${i}`} />
                    ))}
                </tr>
            </thead>
            <tbody>
                {children && children({ grid, dimensions })}
                {!children &&
                    grid.map((row: GridRow, rowIndex: number) => {
                        return (
                            <tr
                                key={`row-${rowIndex}`}
                                className={getPositionnablRowClassNames({ rowIndex }, dimensions)}
                            >
                                {row.columns.map((cell: GridCell, cellIndex: number) => {
                                    const cellStyles = styleForCell ? styleForCell(cell, {}) || {} : {};
                                    const classes = getPositionnableCellClassNames(cell, dimensions);
                                    return (
                                        <td
                                            key={`cell-${rowIndex}-${cellIndex}`}
                                            className={`crystallize-grid__cell ${classes}`}
                                            style={cellStyles}
                                            rowSpan={cell.layout.rowspan}
                                            colSpan={cell.layout.colspan}
                                        >
                                            <CellComponent cell={cell} dimensions={dimensions} />
                                        </td>
                                    );
                                })}
                            </tr>
                        );
                    })}
            </tbody>
        </table>
    );
};
