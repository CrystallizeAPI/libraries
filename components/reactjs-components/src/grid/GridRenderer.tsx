import { FunctionComponent } from 'react';
import { CSSGrid } from './CSSGrid';
import { getGridDimensions } from './grid-renderer-utils';
import { RowCol } from './RowCol';
import { Table } from './Table';
import { GridRow, GridCell, GridRendererProps, GridRenderingType } from './types';

export const GridRenderer: FunctionComponent<GridRendererProps> = ({
    cellComponent,
    children,
    grid,
    type = GridRenderingType.Div,
    styleForCell,
    ...props
}) => {
    if (!cellComponent && !children) {
        console.error('@crystallize/grid-renderer: missing Â´cellComponent` or children function');
        return null;
    }
    if (!grid.rows.length) return null;
    const dimensions = getGridDimensions(grid.rows);

    if (type === GridRenderingType.Table) {
        return (
            <Table
                cellComponent={cellComponent}
                grid={grid.rows}
                dimensions={dimensions}
                styleForCell={styleForCell}
                {...props}
            >
                {children}
            </Table>
        );
    }
    if (type === GridRenderingType.RowCol) {
        return (
            <RowCol
                cellComponent={cellComponent}
                grid={grid.rows}
                dimensions={dimensions}
                styleForCell={styleForCell}
                {...props}
            >
                {children}
            </RowCol>
        );
    }

    const cells = grid.rows.reduce<GridCell[]>((memo, row) => memo.concat(row.columns), []);

    return (
        <CSSGrid
            cellComponent={cellComponent}
            cells={cells}
            dimensions={dimensions}
            styleForCell={styleForCell}
            {...props}
        >
            {children}
        </CSSGrid>
    );
};
