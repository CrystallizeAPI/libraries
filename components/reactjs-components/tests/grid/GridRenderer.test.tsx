import { shallow } from 'enzyme';
import { GridRenderer, GridRenderingType } from '../../src/grid';
import { CellComponent, model } from './utils';

describe('GridRenderer', () => {
    it('renders a grid correctly', () => {
        const wrapper = shallow(<GridRenderer grid={model} cellComponent={CellComponent} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('renders a table correctly', () => {
        const wrapper = shallow(
            <GridRenderer grid={model} cellComponent={CellComponent} type={GridRenderingType.Table} />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('passes the `cellComponent` prop through correctly', () => {
        const wrapper = shallow(<GridRenderer grid={model} cellComponent={CellComponent} />);
        expect(wrapper.find('CSSGrid').prop('cellComponent')).toEqual(CellComponent);
    });

    it('passes the children through correctly', () => {
        // The children should be a function as per the prop types
        const mock = jest.fn();
        const wrapper = shallow(
            <GridRenderer grid={model} cellComponent={CellComponent}>
                {mock}
            </GridRenderer>
        );
        expect(wrapper.find('CSSGrid').prop('children')).toEqual(mock);
    });

    it('should allow custom props to be passed through to the grid component', () => {
        const wrapper = shallow(
            <GridRenderer
                className="my-custom-class"
                grid={model}
                cellComponent={CellComponent}
                type={GridRenderingType.Div}
            />
        );
        expect(wrapper.find('CSSGrid').prop('className')).toEqual('my-custom-class');
    });
});
