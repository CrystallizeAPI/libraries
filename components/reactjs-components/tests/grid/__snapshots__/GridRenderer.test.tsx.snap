// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridRenderer renders a grid correctly 1`] = `ShallowWrapper {}`;

exports[`GridRenderer renders a table correctly 1`] = `ShallowWrapper {}`;
